pkg/controllers/proxy/network_services_controller.go

```diff
@@ -735,75 +744,154 @@ func (nsc *NetworkServicesController) publishMetrics(serviceInfoMap serviceInfoM
 		}
 	}()
 
-	ipvsSvcs, err := nsc.ln.ipvsGetServices()
+	ipvsHandle, err := ipvs.New("")
 	if err != nil {
-		return errors.New("Failed to list IPVS services: " + err.Error())
+		klog.Errorf("failed to initialize ipvs handle: %v", err)
+		return
 	}
+	defer ipvsHandle.Close()
 
-	klog.V(1).Info("Publishing IPVS metrics")
-	for _, svc := range serviceInfoMap {
-		var protocol uint16
-		var pushMetric bool
-		var svcVip string
+	ipvsSvcs, err := ipvsHandle.GetServices()
+	if err != nil {
+		klog.Errorf("failed to list IPVS services: %v", err)
+		return
+	}
+
+	type svcMapKey struct {
+		ip       string
+		uPort    uint16
+		protocol uint16
+	}
 
-		protocol = convertSvcProtoToSysCallProto(svc.protocol)
-		for _, ipvsSvc := range ipvsSvcs {
+	serviceMap := map[svcMapKey]*serviceInfo{}
```

Comment:
I like what you did here with the map instead of having the embeded for loops!

However, this `Collect()` function is still a bit large. Let's go ahead and separate this `serviceMap` building into its own function. We could potentially separate out the second `for` loop into its own function as well and then just call both from `Collect()`.